crypto/account.go

```diff
@@ -145,3 +180,189 @@ func LogicSigAddress(lsig types.LogicSig) types.Address {
+	return AppendMultisigToLogicSig(&lsa.Lsig, signer)
+}
+
+// LogicSigAccountFromLogicSig creates a LogicSigAccount from an existing
+// LogicSig object.
+//
+// The parameter signerPublicKey must be present if the LogicSig is delegated
+// and the delegating account is backed by a single private key (i.e. not a
+// multisig account). In this case, signerPublicKey must be the public key of
+// the delegating account. In all other cases, an error will be returned if
+// signerPublicKey is present.
+func LogicSigAccountFromLogicSig(lsig types.LogicSig, signerPublicKey *ed25519.PublicKey) (lsa LogicSigAccount, err error) {
+	hasSig := lsig.Sig != (types.Signature{})
+	hasMsig := !lsig.Msig.Blank()
+
+	if hasSig && hasMsig {
+		err = errLsigTooManySignatures
+		return
+	}
+
+	if hasSig {
+		if signerPublicKey == nil {
+			err = errLsigNoPublicKey
+			return
+		}
+
+		toBeSigned := programToSign(lsig.Logic)
+		valid := ed25519.Verify(*signerPublicKey, toBeSigned, lsig.Sig[:])
+		if !valid {
+			err = errLsigInvalidPublicKey
+			return
+		}
+
+		lsa.Lsig = lsig
+		lsa.SigningKey = make(ed25519.PublicKey, len(*signerPublicKey))
+		copy(lsa.SigningKey, *signerPublicKey)
+		return
+	}
+
+	if signerPublicKey != nil {
+		err = errLsigAccountPublicKeyNotNeeded
+		return
+	}
+
+	lsa.Lsig = lsig
+	return
+}
+
+// IsDelegated returns true if and only if the LogicSig has been delegated to
+// another account with a signature.
+//
+// Note this function only checks for the presence of a delegation signature. To
+// verify the delegation signature, use VerifyLogicSig.
+func (lsa LogicSigAccount) IsDelegated() bool {
+	hasSig := lsa.Lsig.Sig != (types.Signature{})
+	hasMsig := !lsa.Lsig.Msig.Blank()
+	return hasSig || hasMsig
+}
+
+// Address returns the address of this LogicSigAccount.
+//
+// If the LogicSig is delegated to another account, this will return the address
+// of that account.
+//
+// If the LogicSig is not delegated to another account, this will return an
+// escrow address that is the hash of the LogicSig's program code.
+func (lsa LogicSigAccount) Address() (addr types.Address, err error) {
+	hasSig := lsa.Lsig.Sig != (types.Signature{})
+	hasMsig := !lsa.Lsig.Msig.Blank()
+
+	// require at most one sig
+	if hasSig && hasMsig {
+		err = errLsigTooManySignatures
+		return
+	}
+
+	if hasSig {
+		n := copy(addr[:], lsa.SigningKey)
+		if n != ed25519.PublicKeySize {
+			err = fmt.Errorf("Generated public key has length of %d, expected %d", n, ed25519.PublicKeySize)
+		}
+		return
+	}
+
+	if hasMsig {
+		var msigAccount MultisigAccount
+		msigAccount, err = MultisigAccountFromSig(lsa.Lsig.Msig)
+		if err != nil {
+			return
+		}
+		addr, err = msigAccount.Address()
+		return
+	}
+
+	addr = LogicSigAddress(lsa.Lsig)
+	return
+}
+
+// SignTransaction signs a transaction using this LogicSigAccount.
+func (lsa LogicSigAccount) SignTransaction(tx types.Transaction) (txid string, stxBytes []byte, err error) {
```

Comment:
This matches the account object with the sign function used by Java, but I thought the other SDKs use standalone functions.

Maybe this should be `crypto.LogicSigSignTransaction(lsig, tx)`?

Or maybe SignTransaction could be changed to accept the go `crypto.PrivateKey`, and `LogicSig` can implement that interface?

That said, I'm not really familiar with using smart contracts, so maybe this doesn't work.

Reply:
It looks like you have `SignLogicsigTransaction` also, so there are two ways to sign?

Reply:
Looking a little further, there is a `Signer` interface. Maybe there is a way to use that?

I understand that we're implementing Knapsack soon, so figuring out this interface could be really helpful.

Reply:
My intent was to allow the existing `SignLogicsigTransaction` to still work for compatibility, but to provide an alternative way to sign using the new `LogicSigAccount` object instead of a `LogicSig`. Originally I didn't want to introduce a new `SignLogicSigAccountTransaction` because it sounds too similar, but now I think this might be the best compromise.

Reply:
I'd be interested in having our account objects implement a common `Signer` interface (thought maybe we'd want a more specific `TransactionSigner` interface), but that seems like a bigger change that should happen in a different PR. 

Plus, I'm not aware of the details of how knapsack signatures will be used to anticipate how the SDK can support them yet.